### PR TITLE
Data Editor Improvements for v1.3.1

### LINKED ISSUE #516: Allow multiple simultaneous Data Editors
Currently, the Data Editor only allows one instance to be running at one time.  More than one Data Editor can be run simultaneously if we allow them to share a single instance of an Ωedit server.  This will require some changes to the Data Editor server management code.

### LINKED ISSUE #518: Merge Single and Multiple Byte Editing Modes
If a single byte is selected, then the Editor is implicitly in Single Byte Mode.  If multiple bytes are selected, then the Editor is implicitly in Multiple Byte Mode.

In Single Byte Mode, the Edit View should be replaced entirely with the Data View.
In Multiple Byte Mode, we should only see the Edit View and the Data View should not be visible.

### LINKED ISSUE #568: Click upon seperator space in Logical window
It is possible to click on the space inserted between characters in the Logical window. This results in an unmoved and empty Edit widget and if you were previously in the Physical window, the widget remains there...
image
You can see the outline shows that the Logical window was clicked in, but the Edit widget is still in the Physical, but is now empty.

### LINKED ISSUE #572: Compartmentalize dataEditor.svelte Source
# Overview
A major feature within Svelte is the ability to implement _ReactJS-like_ components with props, reactivity, etc...[^1]

daffodil-vscode v1.3.0's Data Editor source file consists of one massive Svelte component. Aside from readability purposes, there are Svelte syntax indexing issues with having a Svelte file this large. There are also enhancements that could be made with creating individual component files that can utilize onMount.

### Possible Main Components

Some possible main components that could come of this compartmentalization would be components for each main section of the UI:

- DataEditor Header
- Viewports- Configurable per specific viewport through props.
- Data View

While also implementing components for commonly used UI items: buttons, inputs, selects, etc...

[^1]: See https://svelte.dev/docs#component-format for more information.

### LINKED ISSUE #576: Handle changes made to an edited file done outside the editor instance
If changes are made to the file being edited, clear the changes and reset the editor to the new file contents if the file still exists, or close the editor otherwise.

### LINKED ISSUE #631: Add "Save As" to the DataEditor
Currently the DataEditor has a "Save" button which is _really_ "Save As", so this is really to rename the "Save" button to "Save As" and add a "Save" button that overwrites the original file with any in-flight changes.  This will require Ωedit updates to handle the original file being overwritten so in-flight changes don't get cleared and that it gracefully handles changes to the original file that occurred in an external program while the file was being edited in an Ωedit session.

### LINKED ISSUE #654: Allow the DataView to be editable
When a single byte is selected, the DataView appears that shows the byte in Latin-1, and its value in several integer formats.  Currently this is a read-only data display only.  This ticket is to make it so that those values can be edited directly and saved in the same format as displayed.

### LINKED ISSUE #657: Basic data profiling capability
Integrate basic data profiling to show byte frequency graph with an option to download the frequency data in CSV form, and a few common statistical metrics.

### LINKED ISSUE #668: Create Svelte Header Component
### Issue
Currently, the _dataEditor.svelte_ component initializes the `<Main />` and `<ServerMetrics />` via defined Svelte components but the intended Header section is still _inlined_. [^1]

The Header section needs to be broken out into another `<Header />` component.

### Expected Outcome
**dataEditor.svelte**
```ts
<body>
    <Header ... />
    <Main ... />
    <ServerMetrics />
</body>
```

[^1]: In the implement-highlevel-svelte-components branch: https://github.com/ctc-oss/daffodil-vscode/tree/implement-highlevel-svelte-components

### LINKED ISSUE #714: Handle large scale search and replace
Now that we can load and edit very large files, we'll need to also handle searching resulting in over 1K matches, and replacing over 1K matches.

As a test I created a 1MB file with only the character 'a' in it and tried to do a search.  This results in 1M matches, which took only 0.1s.  Next I created a file 100M lines with 'CDE' in every line.  The file is 1.6BG and the number of matches are 100M, this takes way too long (I killed the test after running it for a day).

Since search is having scale issues, replace is worse.  So in addition to having to manage search, replace is tracking and materializing all of the changes on the fly, so 1M matches results in tracking 2M changes in 1M transactions (the delete of the original token and the insertion of the new token).  This took about 3 hours to do, but could take closer to about 0.1s if we did it as a single transaction instead of tracking and materializing all of those individual changes on the fly.

If we don't do the large scale replace in a single transaction, then undo really becomes effectively worthless if the user made some changes, then have 1M transactional changes, going back to the changes made before running the replacements has been effectively disabled.  I view doing the 1M transactions in a single transaction a much more appealing solution for the UI anyway.

### LINKED ISSUE #715: Impose practical limits on search and replace
Now that we open, view, and edit handle very large files, there are challenges with search and replace that won't be solved in time for v1.3.1 (see #714).  In the mean time, create a practical limit for search and replace.

Start with trying 1K and search for up to 1001 matches, then we have either 1K or under matches (to indicate that we have all matches), or 1K+ matches (meaning we have only the _first_ 1K matches, but there are at least 1 more).

This number can be configurable in launch.json. Whatever the number is, add 1 more to it to signal if we have exactly that number of matches or we only have the first set of matches and there are more after this.

### PR DESCRIPTION
**Pull request by stricklandrbls**
- Comparmentalized the dataEditor.svelte into Svelte components.
- Implemented byte edits via the Data View.
- Revised file retention methods.
- Byte editing mode now determined by selection data.
- Implemented overwrite only editing restriction.
- Allow multi instances of the data editor to run at once.
- Implemented basic data profiling panel.
- Transitioned byte display from textarea to byte components.
- Implement scalable search and replace

Closes #715
Closes #714
Closes #668
Closes #657
Closes #654
Closes #631
Closes #576
Closes #568
Closes #518
Closes #516
Closes #572

### CODE REVIEW

**Review by NolanMatt — APPROVED** [review]
Non Binding, but +1

**Review by shanedell — COMMENTED** [review]
Other than the two items commented, everything else looks good. After those two things are resolved I will approve.

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -111,9 +112,15 @@ jobs:
       - run: $SBT compile # used to initialize sbt, if not done sbt universal:packageBin ran by yarn package times out -- windows only
         if: runner.os == 'Windows'
       - run: yarn install
+      - name: Make sure the Ωedit server can start on port 9009
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply .github/workflows/CI.yml:115]: It had problems before, so this was a nice quick sanity check before going full stack.  It's been solid now, so I suppose we can remove this now.

File: .vscode/launch.json [hunk 2]
```diff
@@ -18,6 +18,24 @@
...
+			"infosetOutput": {
+				"type": "file",
+				"path": "${workspaceFolder}/infoset.xml"
+			},
+			"debugServer": 4711,
+			"openHexView": false,
+			"openInfosetView": false,
+			"openInfosetDiffView": false,
+			"daffodilDebugClasspath": ""
+		},
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply .vscode/launch.json:38]: Thanks, I'm glad you saw that.

**Comment by stricklandrbls** [comment]
> Other than the two items commented, everything else looks good. After those two things are resolved I will approve.

@shanedell Can you double check my `ci.yaml` section removal is correct? I've never configured the CI workflows and want to be sure it's as intended.

**Review by shanedell — APPROVED** [review]
+1 👍

**Pull request merged**